File: 2021/python/day10.py
def load_file():
    with open('../inputs/input10') as file:
        return file.read()


test = """[({(<(())[]>[[{[]{<()<>>
[(()[<>])]({[<{<<[]>>(
{([(<{}[<>[]}>{[]{[(<()>
(((({<>}<{<{<>}{[]{[]{}
[[<[([]))<([[{}[[()]]]
[{[{({}]{}}([{[{{{}}([]
{<[[]]>}<{[{[{[]{()[[[]
[<(<(<(<{}))><([]([]()
<{([([[(<>()){}]>(<<{{
<{([{{}}[<[[[<>{}]]]>[]]"""

POINTS_PT1 = {'E': 0, ')': 3, '}': 1197, ']': 57, '>': 25137}
POINTS_PT2 = {'E': 0, ')': 1, '}': 3, ']': 2, '>': 4}
OPEN_CLOSING = {'(': ')', '{': '}', '[': ']', '<': '>'}
CLOSING_OPEN = {')': '(', '}': '{', ']': '[', '>': '<'}


def get_chunks():
    # data = test
    data = load_file()
    return [line for line in data.splitlines()]


def parse(chunk, return_on_error=True):
    """Parse a given chunk and return a tuple with a wrong symbol or 'E' and a list of missing symbols"""
    stack = [chunk[0]]
    i = 1
    error = False
    for c in chunk[1:]:
        if c in OPEN_CLOSING.keys():
            stack.append(c)
        if c in CLOSING_OPEN.keys():
            if len(stack) > 0 and CLOSING_OPEN[c] == stack[len(stack) - 1]:
                stack.pop()
            else:
                error = True
        if error and return_on_error:
            return chunk[i], []
        i += 1

    if not error:
        stack.reverse()
        return 'E', [OPEN_CLOSING[c] for c in stack]
    else:
        return 'E', []


def part1():
    return sum(map(lambda t: POINTS_PT1[t[0]],
                   [parse(chunk) for chunk in get_chunks()]))


def part2():
    results = []
    for added in map(lambda t: t[1], [parse(chunk) for chunk in get_chunks()]):
        points = 0
        for a in added:
            points *= 5
            points += POINTS_PT2[a]
        if points > 0:
            results.append(points)
    results.sort()
    return results[len(results) // 2]


print(part1())  # 26397
print(part2())  # 288957

File: 2021/rust/src/day10.rs
enum ParserOutput {
    Error(char),
    Added(Vec<char>),
}

fn is_opening(c: char) -> bool {
    "([{<".contains(c)
}

fn is_closing(c: char) -> bool {
    ")]}>".contains(c)
}

fn get_closing(c: char) -> Option<char> {
    match c {
        '(' => Some(')'),
        '[' => Some(']'),
        '{' => Some('}'),
        '<' => Some('>'),
        _ => None,
    }
}

fn get_opening(c: char) -> Option<char> {
    match c {
        ')' => Some('('),
        ']' => Some('['),
        '}' => Some('{'),
        '>' => Some('<'),
        _ => None,
    }
}

fn points1(c: char) -> usize {
    match c {
        ')' => 3,
        ']' => 57,
        '}' => 1197,
        '>' => 25137,
        _ => 0,
    }
}

fn points2(c: char) -> usize {
    match c {
        ')' => 1,
        ']' => 2,
        '}' => 3,
        '>' => 4,
        _ => 0,
    }
}

fn parse(chunk: &str) -> ParserOutput {
    let mut stack = vec![];
    let mut i = 0;
    let mut error = false;
    for c in chunk.chars() {
        if is_opening(c) {
            stack.push(c);
        } else if is_closing(c) {
            if !stack.is_empty() && get_opening(c).unwrap() == *stack.last().unwrap() {
                stack.pop().unwrap();
            } else {
                error = true;
            }
        }
        if !error {
            i += 1;
        }
    }

    if !error {
        let mut added = vec![];
        stack.reverse();
        for s in stack {
            added.push(get_closing(s).unwrap());
        }
        ParserOutput::Added(added)
    } else {
        ParserOutput::Error(chunk.chars().nth(i).unwrap())
    }
}

pub fn part1(lines: &Vec<String>) -> usize {
    let mut sum = 0;
    for line in lines {
        sum += match parse(line) {
            ParserOutput::Error(c) => points1(c),
            _ => 0,
        };
    }
    sum
}

pub fn part2(lines: &Vec<String>) -> usize {
    let mut results = vec![];
    for line in lines {
        match parse(line) {
            ParserOutput::Added(added) => {
                let mut points = 0;
                for a in added {
                    points *= 5;
                    points += points2(a);
                }
                if points > 0 {
                    results.push(points);
                }
            }
            _ => {}
        };
    }
    results.sort();
    results[results.len() / 2]
}
